src/library/compiler/ir_interpreter.cpp

```diff
@@ -0,0 +1,759 @@
+    }
+
+    /** \brief Retrieve Lean declaration from environment. */
+    decl get_decl(name const & fn) {
+        option_ref<decl> d = find_env_decl(m_env, fn);
+        if (!d) {
+            throw exception(sstream() << "unknown declaration '" << fn << "'");
+        }
+        return d.get().value();
+    }
+
+    /** \brief Retrieve Lean definition from environment. */
+    decl get_fdecl(name const & fn) {
+        decl d = get_decl(fn);
+        if (decl_tag(d) == decl_kind::Extern) {
+            throw exception(sstream() << "unexpected external declaration '" << fn << "'");
+        }
+        return d;
+    }
+
+    /** \brief Evaluate nullary function ("constant"). */
+    object * load(name const & fn, type t) {
+        if (void * p = lookup_symbol(fn).m_addr) {
+            // constants do not have boxed wrappers, but we'll survive
+            switch (t) {
+                case type::Float: throw exception("floats are not supported yet");
+                case type::UInt8: return box(*static_cast<uint8 *>(p));
+                case type::UInt16: return box(*static_cast<uint16 *>(p));
+                case type::UInt32: return box_uint32(*static_cast<uint32 *>(p));
+                case type::UInt64: return box_uint64(*static_cast<uint64 *>(p));
+                case type::USize: return box_size_t(*static_cast<size_t *>(p));
+                case type::Object:
+                case type::TObject:
+                    return *static_cast<object **>(p);
+                default:
+                    throw exception("invalid type");
+            }
+        } else {
+            push_frame(fn, m_arg_stack.size());
+            decl d = get_fdecl(fn);
+            object * r = eval_body(decl_fun_body(d));
+            pop_frame(r);
+            return r;
+        }
+    }
+
+    object * call(name const & fn, array_ref<arg> const & args) {
+        size_t old_size = m_arg_stack.size();
+
+        // evaluate args in old stack frame
+        for (const auto & arg : args) {
+            m_arg_stack.push_back(eval_arg(arg));
+        }
+
+        decl d = get_decl(fn);
+        push_frame(fn, old_size);
+        object * r;
+        symbol_cache_entry e = lookup_symbol(fn);
+        if (e.m_addr) {
+            if (e.m_boxed) {
+                // NOTE: If we chose the boxed version where the IR chose the unboxed one, we need to manually increment
+                // originally borrowed parameters because the wrapper will decrement these after the call.
+                // Basically the wrapper is more homogeneous (removing both boxed and borrowed parameters) than we
+                // would need in this instance.
+                for (size_t i = 0; i < args.size(); i++) {
+                    if (param_borrow(decl_params(d)[i])) {
+                        inc(m_arg_stack[old_size + i]);
+                    }
+                }
+            }
+            // HACK: `curry` wants a closure object instead of just a function pointer
```

Comment:
This will be fixed, right?

Reply:
Yes. I was trying to adjust `apply.lean` but its syntax is severely out of date... :)

Reply:
Fixed. I edited `apply.cpp` manually because we are missing primitives like `handle.write` for running `apply.lean` right now.

Reply:
This is good. The script is broken, I have been editing `apply.cpp` manually too.